feat: add `ndarray/flatten-by`

headlessNode opened the pull request:
Resolves https://github.com/stdlib-js/metr-issue-tracker/issues/76.

## Description

> What is the purpose of this pull request?

This pull request:

-   add `ndarray/flatten-by`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves https://github.com/stdlib-js/metr-issue-tracker/issues/76
-   progresses #2656

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

headlessNode committed:
feat: add ndarray/flatten-by

headlessNode committed:
refactor: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode commented on lib/node_modules/@stdlib/ndarray/flatten-by/docs/types/test.ts:
```diff
@@ -0,0 +1,150 @@
+/*
+* @license Apache-2.0
+*
+* Copyright (c) 2024 The Stdlib Authors.
```
```suggestion
* Copyright (c) 2025 The Stdlib Authors.
```

headlessNode committed:
docs: apply review suggestion

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode commented on lib/node_modules/@stdlib/ndarray/flatten-by/README.md:
```diff
@@ -0,0 +1,229 @@
+var x = array( [ [ [ 1.0, 2.0 ] ], [ [ 3.0, 4.0 ] ], [ [ 5.0, 6.0 ] ] ] );
+// returns <ndarray>
+
+var y = flatten( x, scale );
```
```suggestion
var y = flattenBy( x, scale );
```

headlessNode commented on lib/node_modules/@stdlib/ndarray/flatten-by/README.md:
```diff
@@ -0,0 +1,229 @@
+
+var x = array( [ [ [ 1.0, 2.0 ] ], [ [ 3.0, 4.0 ] ], [ [ 5.0, 6.0 ] ] ] );
+// returns <ndarray>
+
+var y = flatten( x, {
+    'depth': 1
+});
+// returns <ndarray>
```
```suggestion

var opts = {
     'depth: 1
};
var y = flatten( x, opts, scale );
// returns <ndarray>
```

headlessNode commented on lib/node_modules/@stdlib/ndarray/flatten-by/README.md:
```diff
@@ -0,0 +1,229 @@
+
+var x = array( [ [ [ 1.0, 2.0 ] ], [ [ 3.0, 4.0 ] ], [ [ 5.0, 6.0 ] ] ] );
+// returns <ndarray>
+
+var y = flatten( x, {
+    'order': 'column-major'
+});
+// returns <ndarray>
```
```suggestion

var opts = {
     'order': 'column-major'
};

var y = flatten( x, opts, scale );
// returns <ndarray>
```

headlessNode commented on lib/node_modules/@stdlib/ndarray/flatten-by/README.md:
```diff
@@ -0,0 +1,229 @@
+    'count': 0
+};
+
+var y = flatten( x, scale, ctx );
```
```suggestion
var y = flattenBy( x, scale, ctx );
```

headlessNode commented on lib/node_modules/@stdlib/ndarray/flatten-by/README.md:
```diff
@@ -0,0 +1,229 @@
+});
+console.log( ndarray2array( x ) );
+
+var y = flatten( x, scale );
```
```suggestion
var y = flattenBy( x, scale );
```

headlessNode committed:
docs: apply suggestions from code review

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode commented on lib/node_modules/@stdlib/ndarray/flatten-by/benchmark/benchmark.js:
```diff
@@ -0,0 +1,323 @@
+		zeros( 'float64', [ 10, 10 ], 'row-major' ),
+		zeros( 'float32', [ 10, 10 ], 'row-major' ),
+		zeros( 'int32', [ 10, 10 ], 'row-major' ),
+		zeros( 'complex128', [ 10, 10 ], 'row-major' ),
```
```suggestion
```

headlessNode commented on lib/node_modules/@stdlib/ndarray/flatten-by/benchmark/benchmark.js:
```diff
@@ -0,0 +1,323 @@
+		zeros( 'float64', [ 10, 10 ], 'row-major' ),
+		zeros( 'float32', [ 10, 10 ], 'row-major' ),
+		zeros( 'int32', [ 10, 10 ], 'row-major' ),
+		zeros( 'complex128', [ 10, 10 ], 'row-major' ),
```
```suggestion
```

headlessNode commented on lib/node_modules/@stdlib/ndarray/flatten-by/benchmark/benchmark.js:
```diff
@@ -0,0 +1,323 @@
+		zeros( 'float64', [ 2, 5, 10 ], 'row-major' ),
+		zeros( 'float32', [ 2, 5, 10 ], 'row-major' ),
+		zeros( 'int32', [ 2, 5, 10 ], 'row-major' ),
+		zeros( 'complex128', [ 2, 5, 10 ], 'row-major' ),
```
```suggestion
```

headlessNode commented on lib/node_modules/@stdlib/ndarray/flatten-by/benchmark/benchmark.js:
```diff
@@ -0,0 +1,323 @@
+		zeros( 'float64', [ 2, 5, 10 ], 'row-major' ),
+		zeros( 'float32', [ 2, 5, 10 ], 'row-major' ),
+		zeros( 'int32', [ 2, 5, 10 ], 'row-major' ),
+		zeros( 'complex128', [ 2, 5, 10 ], 'row-major' ),
```
```suggestion
```

headlessNode commented on lib/node_modules/@stdlib/ndarray/flatten-by/benchmark/benchmark.js:
```diff
@@ -0,0 +1,323 @@
+		zeros( 'float64', [ 2, 5, 2, 5 ], 'row-major' ),
+		zeros( 'float32', [ 2, 5, 2, 5 ], 'row-major' ),
+		zeros( 'int32', [ 2, 5, 2, 5 ], 'row-major' ),
+		zeros( 'complex128', [ 2, 5, 2, 5 ], 'row-major' ),
```
```suggestion
```

headlessNode commented on lib/node_modules/@stdlib/ndarray/flatten-by/benchmark/benchmark.js:
```diff
@@ -0,0 +1,323 @@
+		zeros( 'float64', [ 2, 5, 2, 5 ], 'row-major' ),
+		zeros( 'float32', [ 2, 5, 2, 5 ], 'row-major' ),
+		zeros( 'int32', [ 2, 5, 2, 5 ], 'row-major' ),
+		zeros( 'complex128', [ 2, 5, 2, 5 ], 'row-major' ),
```
```suggestion
```

headlessNode commented on lib/node_modules/@stdlib/ndarray/flatten-by/benchmark/benchmark.js:
```diff
@@ -0,0 +1,323 @@
+		zeros( 'float64', [ 2, 5, 2, 5, 1 ], 'row-major' ),
+		zeros( 'float32', [ 2, 5, 2, 5, 1 ], 'row-major' ),
+		zeros( 'int32', [ 2, 5, 2, 5, 1 ], 'row-major' ),
+		zeros( 'complex128', [ 2, 5, 2, 5, 1 ], 'row-major' ),
```
```suggestion
```

headlessNode commented on lib/node_modules/@stdlib/ndarray/flatten-by/benchmark/benchmark.js:
```diff
@@ -0,0 +1,323 @@
+		zeros( 'float64', [ 2, 5, 2, 5, 1 ], 'row-major' ),
+		zeros( 'float32', [ 2, 5, 2, 5, 1 ], 'row-major' ),
+		zeros( 'int32', [ 2, 5, 2, 5, 1 ], 'row-major' ),
+		zeros( 'complex128', [ 2, 5, 2, 5, 1 ], 'row-major' ),
```
```suggestion
```

headlessNode committed:
refactor: apply suggestions from code review

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode committed:
docs: apply suggestions from code review

headlessNode committed:
docs: apply suggestions from code review

headlessNode committed:
docs: apply suggestions from code review

headlessNode commented on lib/node_modules/@stdlib/ndarray/flatten-by/README.md:
```diff
@@ -0,0 +1,233 @@
+// returns <ndarray>
+
+var arr = ndarray2array( y );
+// returns [ 2.0, 6.0, 4.0, 10.0, 8.0, 12.0 ]
```
```suggestion
// returns [ 2.0, 6.0, 10.0, 4.0, 8.0, 12.0 ]
```

headlessNode committed:
docs: apply suggestions from code review

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

kgryte commented on lib/node_modules/@stdlib/ndarray/flatten-by/README.md:
```diff
@@ -0,0 +1,233 @@
+
+# flattenBy
+
+> Apply a callback to elements in an input [ndarray][@stdlib/ndarray/ctor] and return a flattened output [ndarray][@stdlib/ndarray/ctor].
```
@headlessNode See the description in https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/array/base/flatten-by.

kgryte commented on lib/node_modules/@stdlib/ndarray/flatten-by/lib/main.js:
```diff
@@ -0,0 +1,106 @@
+	if ( !isFunction( cb ) ) {
+		throw new TypeError( format( 'invalid argument. Callback argument must be a function. Value: `%s`.', cb ) );
+	}
+	y = map( x, cb, ctx );
+	return flatten( y, opts );
```
You need to rethink this, as you are creating what could be a large temporary array. There is likely a more efficient way. Namely, similar to flatten, allocate a contiguous output array with the desired order, create a view over the output array which matches the input array, and then call ndarray/base/map. Then, you should be able to transform and flatten in a single call.

kgryte replied:
One thing we should note in the docs is that array traversal order is dependent on the memory layout of the input and output arrays. Meaning, it is not guaranteed that a provided callback will be invoked with elements in view order. That is likely fine for 99% of use cases, but users should not make any assumptions regarding traversal order.

headlessNode committed:
refactor: apply suggestions from code review

headlessNode committed:
docs: addd note

headlessNode committed:
refactor: apply suggestions from code review

headlessNode committed:
test: add missing test cases

headlessNode committed:
docs: remove spaces

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
fix: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
fix: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

headlessNode commented:
@kgryte is this a new lint rule to add trailing `[]` on reference links in `.md`?

headlessNode committed:
fix: lint error

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/ndarray/flatten-by/README.md:
````diff
@@ -0,0 +1,235 @@
+var flattenBy = require( '@stdlib/ndarray/flatten-by' );
+```
+
+#### flattenBy( x\[, options], fcn[, thisArg] )
````
```suggestion
#### flattenBy( x\[, options], fcn\[, thisArg] )
```

kgryte commented:
No, that is a legitimate error. You are likely not properly escaping.

headlessNode committed:
test: add tests

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
test: add tests

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
test: add tests

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
test: add test case

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
chore: clean-up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

kgryte reviewed (approved):
LGTM after clean-up.

@headlessNode You'll want to take a look at how I simplified the TS declarations.

As discussed on Slack, a follow-up PR should add `dtype` option support.
